syntax = "proto3";

package banking;

service TransactionService {
  // Transaction operations
  rpc ProcessTransaction (ProcessTransactionRequest) returns (Transaction);
  rpc GetTransaction (GetTransactionRequest) returns (Transaction);
  rpc ListTransactions (ListTransactionsRequest) returns (ListTransactionsResponse);
  
  // Specialized transaction types
  rpc ProcessDeposit (DepositRequest) returns (Transaction);
  rpc ProcessWithdrawal (WithdrawalRequest) returns (Transaction);
  rpc ProcessTransfer (TransferRequest) returns (Transaction);
}

message ProcessTransactionRequest {
  string user_id = 1;
  string account_id = 2;
  string transaction_type = 3;
  string amount = 4;
  string description = 5;
  string reference_id = 6;
}

message GetTransactionRequest {
  string transaction_id = 1;
}

message ListTransactionsRequest {
  string user_id = 1;
  string account_id = 2;
  string start_date = 3;
  string end_date = 4;
  string transaction_type = 5;
  int32 page = 6;
  int32 page_size = 7;
}

message ListTransactionsResponse {
  repeated Transaction transactions = 1;
  int32 total_count = 2;
}

message DepositRequest {
  string user_id = 1;
  string account_id = 2;
  string amount = 3;
  string description = 4;
  string payment_method = 5;
}

message WithdrawalRequest {
  string user_id = 1;
  string account_id = 2;
  string amount = 3;
  string description = 4;
  string withdrawal_method = 5;
}

message TransferRequest {
  string user_id = 1;
  string from_account_id = 2;
  string to_account_id = 3;
  string amount = 4;
  string description = 5;
}

message Transaction {
  string transaction_id = 1;
  string user_id = 2;
  string account_id = 3;
  string transaction_type = 4;
  string amount = 5;
  string description = 6;
  string date_posted = 7;
  string reference_id = 8;
  string status = 9;
}